Add absolute normalisation to POLARIS powder diffraction scripts

DannyHindson opened the pull request:
**Description of work.**

Add a switch to allow an absolute normalisation to be performed during the POLARIS reduction through the ISIS powder diffraction scripts. This will produce a differential cross section with the correct scale. The change was primarily motivated by the total scattering workflow where a Placzek correction is subtracted and therefore the intensity needs to be of the correct scale.

As part of this the interface of some classes used in the workflow have been updated:

- a switch has been added to the CalculatePlaczekSelfScatteringCorrection v1 algorithm to turn off the part of the calculation where the correction size is scaled by a packing fraction. It's not very clear why this scaling was added but it doesn't appear to be required\appropriate if the intensity has been properly normalized

- the setting "crystal_density" has been removed from the SampleDetails class. The density of the material can now be set instead by supplying two of "number_density", "number_density_effective" and "packing_fraction". These three properties on the SampleDetails class are passed through to the Kernel::Material class where they already existed as a result of some work done by ORNL

Finally, there were a couple of glitches relating to the Kernel::Material class that I've fixed up:

- the MayersSampleCorrection algorithm (which is used in the ISIS Powder diffraction scripts) now uses the powder density rather than the full crystal density
- the logic to set the number_density_effective property in SetSample had a bug in it
- you can now supply a chemical formula for an element (eg Si) and a packing_fraction and Mantid calculates the number density and effective number density from these inputs. Previously an incorrect validation prevented this 

**To test:**

Unzip this zip file somewhere: 
[POLARISPowderDiffractionFolderSmall.zip](https://github.com/mantidproject/mantid/files/9103459/POLARISPowderDiffractionFolderSmall.zip)

Update the paths in the file polaris_config_example.yaml so they match where you unzipped the file and also update config_file_path in the script below

Run the following script in workbench. The script creates various workspaces including a workspace group called s_of_q_corrected. There are 5 single spectrum workspaces in here corresponding to the 5 banks on the POLARIS instrument. They should all have y data that is >=1 at all values of Q as a result of this PR. This workspace gets 1 subtracted from it before being passed as input to the PDFFourierTransform algorithm. So this check means the input to PDFFourierTransform is positive:

```
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

from isis_powder import polaris, SampleDetails

config_file_path = r"C:\Total Scattering\polaris_config_example.yaml"
polaris = polaris.Polaris(config_file=config_file_path, user_name="test", mode="PDF")

sample_details = SampleDetails(height=4.0, radius=0.2985, center=[0, 0, 0], shape='cylinder')
sample_details.set_material(chemical_formula='Si', packing_fraction=0.6)
polaris.set_sample_details(sample=sample_details)

polaris.create_vanadium(first_cycle_run_no="98532", multiple_scattering=False)

polaris.focus(run_number="98533", input_mode='Summed')

q_lim = [[2.5, 3, 4, 6, 7], [3.5, 5, 7, 11, 40]]

polaris.create_total_scattering_pdf(run_number="98533", merge_banks=True,q_lims=q_lim, 
                                    freq_params=[1.4], pdf_type="g(r)", debug=True)
```

Fixes #33606.

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

DannyHindson committed:
Changes to provide absolute normalisation option

Formulae based on Stewart lecture notes from Oxford School on
Neutron Scattering with thickness replaced by volume to cope
with none cuboid sample shapes
Enable for POLARIS only at the moment when mode="PDF"
Also remove default value from the sample_details parameter of
generate_ts_pdf because the function didn't work with this default

DannyHindson committed:
Add some debug output to new normalisation

DannyHindson committed:
Stop scaling Placzek correction by packing fraction

Don't think the Plazcek correction needs to be scaled by a packing
fraction. Have checked with Helen Playford + ORNL don't scale by
a packing fraction in their MTS reduction
In the spirit of not making breaking changes to algorithm parameter
lists I have left the crystal_density parameter in place on the
CalculatePlaczekSelfScattering v1 algorithm. I have removed it from
the surrounding workflow algorithm though on basis that this is only
used by ISIS

DannyHindson committed:
Use effective number density instead of number density

In case someone has set a packing fraction on the material
Not currently supported via the ISIS Powder Script sample_details
class but may add packing fraction support in future - also user
might have set packing fraction by running SetSample directly on
the input ws

DannyHindson committed:
Add way of disabling packing fraction scaling

Inside CalculatePlaczekSelfScattering
Also switch from using a crystal_density parameter in the POLARIS
powder scripts over to using number_density, number_density_effective
and packing fraction as is done in Kernel

DannyHindson committed:
Make MayersSampleCorrection algorithm respect packing fraction

The density used in the absorption correction should respect any
packing fraction set on the material.
Do this by calling numberDensityEffective()
This aligns this correction algorithm with the behaviour of
MonteCarloAbsorption and AbsorptionCorrection

DannyHindson committed:
Add new density sample properties to docs

DannyHindson committed:
Tidy up some problems from rebase

Also finish off plumbing in packing_fraction into the powder
diffraction scripts

DannyHindson committed:
Fix flake8 issues

DannyHindson committed:
Add release note

DannyHindson committed:
Set up normalisation method as its own parameter

Instead of driving it purely from the "mode"
Defaults have been set up that depend on the mode though (PDF vs Rietveld)

DannyHindson committed:
Fix bug in absolute normalisation formula

DannyHindson committed:
Use s rather than z in normalise

DannyHindson committed:
Fix failing unit tests

DannyHindson committed:
Add documentation on abs normalisation flag

DannyHindson committed:
Expand description of ts workflow to include conversion to S(Q)

DannyHindson committed:
Allow density to be specified with packing fraction only

This should be allowed if the material specified is an element
If it's not an element then it shouldn't pass. Since the chemical formula
size is only available in MaterialBuilder.cpp just leave this class to
do that validation

DannyHindson committed:
Update total scattering input data file for POLARIS system tests

Incorporate new absolute normalisation into the input data file
Also include a couple of other changes in the file:
- the TOF cropping values have changed since the file was last updated
  This changes the bin number where the 3.9 Angstroms Si peak sits so
  make the tests lookup the bin number rather than hardcoding it
- the new file was generated with absorption correction applied during
the focus process

DannyHindson committed:
Check full and effective number density in test

Documentation says mass density sets effective number density
so the "copy through" into the full number density that the test
relied on seems shaky. Since the element can be identified by the
supplied atomic number it seems reasonable that the full density
is looked up from the Mantid tables
Also update the documentation because case where number density
looked for an element wasn't mentioned

DanielMurphy22 reviewed (dismissed):
Edit: This is already a separate issue https://github.com/mantidproject/mantid/issues/33552

I think I followed your instructions correctly and I get this error:

```
The two workspaces are not compatible because they have different units for the data (Y).
The two workspaces are not compatible for algorithm Minus
Error in execution of algorithm Minus:
The two workspaces are not compatible for algorithm Minus
RuntimeError: Minus-v1: Execution of Minus for group entry 1 failed: The two workspaces are not compatible for algorithm Minus
  File "/home/danielmurphy/Downloads/DannyTest/DannyTest.py", line 21, in <module>
    polaris.create_total_scattering_pdf(run_number="98533", merge_banks=True,q_lims=q_lim,
  File "/home/danielmurphy/mantid/scripts/Diffraction/isis_powder/polaris.py", line 62, in create_total_scattering_pdf
    pdf_output = polaris_algs.generate_ts_pdf(run_number=self._inst_settings.run_number,
  File "/home/danielmurphy/mantid/scripts/Diffraction/isis_powder/polaris_routines/polaris_algs.py", line 120, in generate_ts_pdf
    focused_ws = mantid.Subtract(LHSWorkspace=focused_ws, RHSWorkspace=self_scattering_correction)
  File "/home/danielmurphy/mantid/Framework/PythonInterface/mantid/simpleapi.py", line 1065, in __call__
    raise RuntimeError(msg) from e
```

It looks like the workspaces `focused_ws_1-5` are not in the group `focused_ws` in the ADS

DanielMurphy22 commented on docs/source/release/v6.5.0/Diffraction/Powder/New_features/34193 Absolute Normalization.txt:
```diff
@@ -0,0 +1 @@
+- Add a switch to allow an absolute normalization to be performed during by the ISIS powder diffraction scripts when running on the POLARIS instrument. As part of this the setting "crystal_density" has been removed from the SampleDetails class. The density of the material can now be set instead by supplying two of "number_density", "number_density_effective" and "packing_fraction"
```
```
during by the ISIS powder diffraction ...
```
Just reword slightly. Also should these release note files be .rst??? Maybe it doesn't matter!

DanielMurphy22 commented on docs/source/techniques/ISISPowder-SampleDetails-v1.rst:
```diff
@@ -251,28 +252,47 @@ This parameter defines the number density of the property.
+
+packing_fraction
+^^^^^^^^^^^^^^^^
+This is an alternative to using number_density_effective to specify a powder_density
```
:ref:`number_density_effective_sampleDetails_isis-powder-diffraction-ref`

DanielMurphy22 commented on docs/source/techniques/ISISPowder-Polaris-v1.rst:
```diff
@@ -402,6 +412,30 @@ Example Input:
+
+If "Absolute" is selected then the measured intensity is multipled by the following additional factor:
+
+.. math::
+
+    \frac{\rho_v V_v \sigma_v}{4 \pi \rho_s V_s}
```
Is it worth adding the reference for this?? Feel free to say no!

DannyHindson replied:
Yes I'll put one in from the Howe paper that we've been occasionally referring to in our discussions with Helen

DanielMurphy22 reviewed (changes requested):
Just a few docs things. Really good otherwise! Good job spotting and fixing the 3 bugs along the way!

DannyHindson committed:
Various documentation improvements

DanielMurphy22 reviewed (dismissed):
Looks good to me!

DannyHindson committed:
Rework the conversion from a diff cross section to S(Q)

Subtract the Placzek correction from the dcs rather than S(Q)
- the S(Q) referred to in the Howe paper is closer to a diff cross
  section than what is commonly referred to as S(Q)
Use different formula to convert from diff cross section to S(Q)
so that any incoherent scattering is removed. The new formula also
achieves the same effect as subtracting one from S(Q) so no need to
do that explicitly any more

DannyHindson committed:
Few minor tidy ups

Hide the abs normalisation factor workspace
Adjust S(Q)-1 documentation so it's more clear

DannyHindson commented:
@DanielMurphy22  - I've finished digesting the email conversation with Helen about the latest formula in generate_ts_pdf where I convert a differential cross section into S(Q). I'd already made the changes but hadn't pushed them until now and am happy they're correct. So please can you retest when you have a bit of time

DannyHindson commented:
Latest formula to convert from differential cross section to "S(Q)-1" is in the attached paper. If you substitute equation 11 into equation 13 you'll get an expression for "S(Q)-1" that looks like what I've done in the most recent two commits:

[j-54-00317.pdf](https://github.com/mantidproject/mantid/files/9190194/j-54-00317.pdf)

DannyHindson committed:
Update system tests results following change to S(Q)-1 formula

DanielMurphy22 reviewed (changes requested):
Functionally looks good to me! Just one docs question!

DanielMurphy22 commented:
Actually, in the case that the data was focused with vanadium norm method Relative, should create_total_scattering_pdf() throw an error?

DannyHindson committed:
Align S(Q)-1 formula in docs with code

Also replace < > with special \langle and \rangle symbols

DannyHindson commented on docs/source/techniques/ISISPowder-Polaris-v1.rst:
```diff
@@ -167,21 +168,33 @@ create_total_scattering_pdf
+- subtracts the Placzek self scattering correction from the differential cross section
+- converts the differential cross-section into an :math:`S(Q) - 1` distribution using the following relationship where :math:`\sigma_s` is the total scattering cross section and :math:`\sigma_{coh}` is the coherent scattering cross section. This is based on equations 5, 9 and 19 in [#Keen]_:
+
+:math:`\hspace{20pt}S(Q) - 1 = (\frac{d\sigma_s}{d\Omega}(Q) - <b_{tot}^2>) / <b_{coh}>^2 = (\frac{d\sigma_s}{d\Omega}(Q) - \frac{\sigma_s}{4 \pi}) / <b_{coh}>^2`
```
I've just pushed a commit and I went for your suggested change in the end (rather than changing the code to match the documentation). This was partly because it seems cleaner having the equation all in terms of cross sections instead of a mix of one cross section and one scattering length. Also the cross sections are visible to the user in the UI on the right click "Show Sample Material" menu on a workspace so perhaps these quantities are a bit more user friendly:

![image](https://user-images.githubusercontent.com/56295817/181452835-764ed240-2d4c-4e1b-96a3-7e949311dfe2.png)

DannyHindson commented:
> Actually, in the case that the data was focused with vanadium norm method Relative, should create_total_scattering_pdf() throw an error?

I just had a look at implementing this and it's not very easy. The user can specify the normalisation method in the parameters of the focus call in a script as well as in the .yaml file. So it's not very easy to work out what the user actually used on the focus call prior to calling create_total_scattering_pdf. In the similar change Sarah did recently (https://github.com/mantidproject/mantid/pull/34224) she was able to check whether vanadium normalisation had been done at all by checking whether the workspace output by focus was a distribution or not. The only difference between the relative and absolute normalisation is the size of the y values.

So the check you're suggesting is something we'd do in an ideal world but since it's not trivial and the potential for it happening is not that high I think it's OK to leave for now. We've previously discussed with Helen adding a feature to tune the packing_fraction to that S(Q) tends to 1 at large Q so perhaps we could incorporate a warning if the scale factor required to do this is way off. That might be the most clear signal that the wrong type of normalisation has been done.

DanielMurphy22 reviewed (changes requested):
<img width="759" alt="Screenshot 2022-07-28 at 10 29 03" src="https://user-images.githubusercontent.com/55980573/181472346-faf1176e-f780-45a2-851b-65a3c7cdebb6.png">

Sorry, the docs are very close. I think the second b is missing the `coh` subscript! And oddly one of the overlines hasn't been generated. Is this the case on your machine?

DannyHindson commented:
> <img alt="Screenshot 2022-07-28 at 10 29 03" width="759" src="https://user-images.githubusercontent.com/55980573/181472346-faf1176e-f780-45a2-851b-65a3c7cdebb6.png">
> 
> Sorry, the docs are very close. I think the second b is missing the `coh` subscript! And oddly one of the overlines hasn't been generated. Is this the case on your machine?

Actually $b_{coh}=\overline{b}$ so I think it's OK. So $\langle b_{coh} \rangle^2 = \langle \overline{b} \rangle^2$ . I'd changed the notation on the b's used in the middle bit of the equation because it aligns more with the Keen paper that Helen likes.

DanielMurphy22 commented:
> > <img alt="Screenshot 2022-07-28 at 10 29 03" width="759" src="https://user-images.githubusercontent.com/55980573/181472346-faf1176e-f780-45a2-851b-65a3c7cdebb6.png">
> > Sorry, the docs are very close. I think the second b is missing the `coh` subscript! And oddly one of the overlines hasn't been generated. Is this the case on your machine?
> 
> Actually bcoh=b― so I think it's OK. So ⟨bcoh⟩2=⟨b―⟩2 . I'd changed the notation on the b's used in the middle bit of the equation because it aligns more with the Keen paper that Helen likes.
> 
> Re the overlines - on my machine it looks like this:
> 
> ![image](https://user-images.githubusercontent.com/56295817/181480082-dd1aed93-bb08-43fa-b6bc-e34845aa0e29.png)
> 
> So yes one of the overlines looks like it's gone. My screenshot is from the html docs. How about yours - was that also from the html docs or from the workbench docs?

It's from my html docs, but happy if it works on your machine. Do you need to change from the b^coh in the text below the equation to align with the new overline system?

DannyHindson commented:
Strange the overline behaviour is different for you. I guess you're on a Conda set up on Mac? Yes you're right the text below doesn't line up any more. I'll just remove $b_{coh}$ from the definitions sentence below the equation

DanielMurphy22 commented:
> I guess you're on a Conda set up on Mac? 

Conda on Linux so we should check that the online docs have the overline when this is merged, as I think they are generated on Rhel7.

DanielMurphy22 reviewed (approved):
Thanks for the recent docs changes! All looks good to me!
